### PR TITLE
[Feature][TASK-PLUGIN] AWS Database Migration Service (DMS)

### LINKED ISSUE #11672: [Feature][TASK-PLUGIN] AWS Database Migration Service (DMS)
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

AWS Database Migration Service (AWS DMS) helps you migrate databases to AWS quickly and securely. 

https://aws.amazon.com/dms

Create a new task type to deal with AWS Database Migration Service Operator.
Maybe support:
- [ ] Create and start replication tasks to migrate databases.

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by jieguangzhou**
## Purpose of the pull request

close: #11672

## Brief change log

- [x] Create and start DMS task by interface
- [x] Create and start DMS task by json data
- [x] Restart DMS task by interface
- [x] Restart DMS task by json data

**Create and start DMS task by interface**

image

**Create and start DMS task by json data**

image

**Restart DMS task by interface**

image

**Restart DMS task by json data**

image

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

### CODE REVIEW

**Comment by caishunfeng** [comment]
Hi @jieguangzhou is this pr ready for review?

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/pom.xml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        <dependency>
+            <groupId>org.apache.dolphinscheduler</groupId>
+            <artifactId>dolphinscheduler-common</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.amazonaws</groupId>
+            <artifactId>aws-java-sdk-dms</artifactId>
+            <version>1.12.297</version>
+        </dependency>
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/pom.xml]: Move the version to bom.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHook.java [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+                }
+                if (!lastPercent.equals(percent)) {
+                    String runningMessage = String.format("fullLoadProgressPercent: %s ", percent);
+                    logger.info(runningMessage);
+                }
+                lastPercent = percent;
+            }
+
+            if (exceptStatus.equals(status)) {
+                logger.info("success");
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHook.java:261]: remove it if useless.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHook.java [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+                return true;
+            } else if (stopStatusSet.contains(status)) {
+                break;
+            }
+        }
+        logger.info("error");
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHook.java:267]: remove it if useless.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTask.java [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+    @Override
+    public DmsParameters getParameters() {
+        return parameters;
+    }
+
+    @Override
+    public void cancelApplication() {
+        dmsHook.stopReplicationTask();
+//        dmsHook.deleteReplicationTask();
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTask.java]:
```suggestion
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsParameters.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    private String replicationTaskArn;
+    private String startReplicationTaskType;
+
+    @Override
+    public boolean checkParameters() {
+        boolean flag;
+        if (isJsonFormat) {
+            flag = jsonData != null;
+        } else if (isRestartTask) {
+            flag = (replicationTaskArn != null);
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsParameters.java]:
```suggestion
            flag = replicationTaskArn != null;
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTask.java [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+        if (parameters.getIsJsonFormat() && parameters.getJsonData() != null) {
+            // combining local and global parameters
+            String jsonData = ParameterUtils.convertParameterPlaceholders(parameters.getJsonData(), ParamUtils.convert(taskExecutionContext.getPrepareParamsMap()));
+
+            boolean isRestartTask = parameters.getIsRestartTask();
+            try {
+                parameters = objectMapper.readValue(jsonData, DmsParameters.class);
+                parameters.setIsRestartTask(isRestartTask);
+            } catch (Exception e) {
+                logger.error("Failed to convert json data to DmsParameters object, error message: {}", e.getMessage());
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTask.java]:
```suggestion
                logger.error("Failed to convert json data to DmsParameters object", e);
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHookTest.java [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+    @Test(timeout = 60000)
+    public void testCreateReplicationTask() throws Exception {
+
+        DmsHook dmsHook = spy(new DmsHook());
+        CreateReplicationTaskResult createReplicationTaskResult = mock(CreateReplicationTaskResult.class);
+        when(client.createReplicationTask(any())).thenReturn(createReplicationTaskResult);
+
+        ReplicationTask replicationTask = mock(ReplicationTask.class);
+        when(replicationTask.getReplicationTaskArn()).thenReturn("arn:aws:dms:ap-southeast-1:123456789012:task:task");
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHookTest.java]: use constant string `arn:aws:dms:ap-southeast-1:123456789012:task:task`

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHookTest.java [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+        String pathParameter = "file://" + path;
+        Assert.assertEquals(jsonData, dmsHook.replaceFileParameters(pathParameter));
+
+//        String pathParameter2 = "file://" + "not_exist.json";
+//
+//        try {
+//            Assert.assertEquals(pathParameter2, dmsHook.replaceFileParameters(pathParameter2));
+//        }catch (Exception e) {
+//            Assert.assertTrue(e instanceof IOException);
+//        }
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHookTest.java]:
```suggestion
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHookTest.java [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+
+        String pathParameter3 = "{}";
+        Assert.assertEquals(pathParameter3, dmsHook.replaceFileParameters(pathParameter3));
+
+    }
+
+//    this.getClass().getResourceAsStream("SagemakerRequestJson.json"))
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHookTest.java]:
```suggestion
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTaskTest.java [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    public void before() throws Exception {
+        whenNew(DmsHook.class).withAnyArguments().thenReturn(dmsHook);
+        DmsParameters dmsParameters = new DmsParameters();
+        dmsTask = initTask(dmsParameters);
+        dmsTask.initDmsHook();
+        MemberModifier.field(DmsTask.class, "dmsHook").set(dmsTask, dmsHook);
+    }
+
+    @Test
+    public void testCreateTaskJson() {
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTaskTest.java:57]: It seems a meanless test case.

Comment (jieguangzhou) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTaskTest.java:57]: This case will make sure that the parameter can read value correctly from JSON format data

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHook.java (hunk 2 above)
Comment (jieguangzhou) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHook.java:261]: I think It is necessary to log messages about the DMS task connect status for every step.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHook.java (hunk 3 above)
Comment (jieguangzhou) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHook.java:267]: I think It is necessary to log messages about the DMS task connect status for every step. Especially when it's wrong.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHookTest.java [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.dolphinscheduler.plugin.task.dms;
+
+import static org.mockito.Mockito.any;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.spy;
+import static org.powermock.api.mockito.PowerMockito.mock;
```
Comment (EricGao888) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHookTest.java]: Could u plz removing the dependency of `powermock` in this PR? We have already had a PR to remove all `powermock` in `dolphinscheduler-task-plugin` module, see: https://github.com/apache/dolphinscheduler/pull/11778
Thanks~

**Review by EricGao888 — CHANGES_REQUESTED** [review]

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHookTest.java (hunk 11 above)
Comment (jieguangzhou) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHookTest.java]: Yes, I will remove powermock

Comment (jieguangzhou) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/test/java/org/apache/dolphinscheduler/plugin/task/dms/DmsHookTest.java]: done

**Comment by jieguangzhou** [comment]
@caishunfeng @EricGao888 PTAL, thanks.

File: docs/docs/en/guide/task/dms.md [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+DMS task plugin can help users to create and start DMS tasks in DolphinScheduler more conveniently.
+
+Contains two features:
+- Create DMS task and start DMS task
+- Restart DMS task
+
+We can create DMS task and start DMS task in two ways:
+- Use interface
+- Use json data
```
Comment (Tianqi-Dotes) [inline docs/docs/en/guide/task/dms.md]: JSON

File: docs/docs/en/guide/task/dms.md [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+DolphinScheduler will track the status of the DMS task and set the status to successfully completed when the DMS task is completed. Except for the CDC task without end time.
+
+So, if the `migrationType` is `cdc` or `full-load-and-cdc`, `cdcStopPosition` not be set, DolphinScheduler will set the status to successfully after the DMS task start successfully.
```
Comment (Tianqi-Dotes) [inline docs/docs/en/guide/task/dms.md]: after the DMS task starts successfully.

File: docs/docs/en/guide/task/dms.md (hunk 12 above)
Comment (Tianqi-Dotes) [reply docs/docs/en/guide/task/dms.md]: and the following json

File: docs/docs/en/guide/task/dms.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+
+### First, introduce some general parameters of DolphinScheduler
+
+- **Node name**: The node name in a workflow definition is unique.
+- **Run flag**: Identifies whether this node schedules normally, if it does not need to execute, select
+  the `prohibition execution`.
+- **Descriptive information**: Describe the function of the node.
+- **Task priority**: When the number of worker threads is insufficient, execute in the order of priority from high
+  to low, and tasks with the same priority will execute in a first-in first-out order.
```
Comment (Tianqi-Dotes) [inline docs/docs/en/guide/task/dms.md]: first-in-first-out

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTask.java [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+     * init DMS remote AppId if null
+     */
+    private void initAppId() {
+        if (appId == null) {
+            if (StringUtils.isNotEmpty(getAppIds())) {
+                appId = JSONUtils.parseObject(getAppIds(), DmsHook.ApplicationIds.class);
+            }
+        }
+        if (appId == null) {
+            throw new TaskException("sagemaker applicationID is null");
```
Comment (Tianqi-Dotes) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTask.java]: sagemaker

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTask.java [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+        return Collections.emptyList();
+    }
+
+    @Override
+    public void submitApplication() throws TaskException {
+        exitStatusCode = checkCreateReplicationTask();
+        if (exitStatusCode == TaskConstants.EXIT_CODE_SUCCESS) {
+            exitStatusCode = startReplicationTask();
+        } else {
+            throw new TaskException("DMS task failed to start");
```
Comment (Tianqi-Dotes) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTask.java]: this else haven't start task yet

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTask.java [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+        }
+
+        // if the task is not running, the task will be deleted
+        if (exitStatusCode == TaskConstants.EXIT_CODE_FAILURE && !parameters.getIsRestartTask()) {
+            dmsHook.deleteReplicationTask();
+        }else {
```
Comment (Tianqi-Dotes) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTask.java]: format

File: dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTask.java [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+    public void initDmsHook() throws TaskException {
+        convertJsonParameters();
+
+        dmsHook = new DmsHook();
+        try {
+            BeanUtils.copyProperties(dmsHook, parameters);
+        } catch (Exception e) {
+            throw new TaskException("DMS task init error", e);
+        }
+
```
Comment (Tianqi-Dotes) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-dms/src/main/java/org/apache/dolphinscheduler/plugin/task/dms/DmsTask.java:217]: format

**Review by Tianqi-Dotes — COMMENTED** [review]
PTAL

**Comment by jieguangzhou** [comment]
> PTAL

Done, PTAL, thanks

**Review by caishunfeng — DISMISSED** [review]
LGTM

File: docs/docs/en/guide/task/dms.md [hunk 19]
```diff
@@ -0,0 +1,107 @@
+# DMS Node
+
+## Overview
+
+[Pytorch](https://pytorch.org) is a mainstream Python machine learning library.
```
Comment (EricGao888) [inline docs/docs/en/guide/task/dms.md]: Why there is a link to `pytorch` here? Maybe we could remove it.

File: docs/docs/en/guide/task/dms.md [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+
+![dms](../../../../img/tasks/demo/dms_create_and_start_json.png)
+
+**Restart DMS task by JSON data**
+
+![dms](../../../../img/tasks/demo/dms_restart_json.png)
+
+
+
+### First, introduce some general parameters of DolphinScheduler
```
Comment (EricGao888) [inline docs/docs/en/guide/task/dms.md:51]: We have put those default parameters in a separate doc for better maintainability. See: https://github.com/apache/dolphinscheduler/pull/11776

**Review by EricGao888 — DISMISSED** [review]
LGTM as long as docs fixed

File: docs/docs/en/guide/task/dms.md (hunk 19 above)
Comment (jieguangzhou) [reply docs/docs/en/guide/task/dms.md]: done

File: docs/docs/en/guide/task/dms.md (hunk 20 above)
Comment (jieguangzhou) [reply docs/docs/en/guide/task/dms.md:51]: done

**Review by EricGao888 — APPROVED** [review]

**Comment by jieguangzhou** [comment]
@Amy0104 PTAL, thanks

**Review by labbomb — APPROVED** [review]
LGTM

**Review by devosend — APPROVED** [review]
LGTM for front-end

**Comment by jieguangzhou** [comment]
Thanks all

**Pull request merged**